keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1294 @@
+`DeleteOptions` is used to delete pods and to modify the code by adding a
+meaningful `reason` field based on the invocation context.
+
+Notably, the `reason` field might be useful to be supplied when deleting other
+types of kubernetes resources, but supplying the field in other contexts is
+beyond this KEP.
+
+In the case of Pods, kube-apiserver will store the received `reason` in the
+already existing `reason` field in the `PodStatus`:
+```
+// PodStatus represents information about the status of a pod. Status may trail the actual
+// state of a system.
+type PodStatus struct {
+  ...
+	// A brief CamelCase message indicating details about why the pod is in this state. e.g. 'Evicted'
+	// +optional
+	Reason string
+  ...
+}
+```
+
+### JobSpec API
+
+We extend the Job API in order to allow to apply different actions depending
+on the conditions associated with the pod failure.
+
+```golang
+// BackoffPolicyAction specifies how a Pod failure is handled.
+type BackoffPolicyAction string
+
+const (
+
+	// This is an action which might be taken on pod failure - mark the
+	// pod's job as Failed and terminate all running pods.
+	TerminateBackoffAction BackoffPolicyAction = "Terminate"
+
+	// This is an action which might be taken on pod failure - the pod will be
+	// restarted and the counter for .backoffLimit will not be incremented.
+	IgnoreBackoffAction BackoffPolicyAction = "Ignore"
+
+)
+
+// ExitCodeInterval specifies an interval of exit codes
+type ExitCodeInterval struct {
+	// specifies the lower end (inclusive) of the interval
+	min int
+	// specifies the upper end (inclusive) of the interval
+	max int
+}
+
+// ExitCodeCondition specifies the condition on the exit code of the failed
+// container
+type ExitCodeCondition struct {
+	// specifies the set of internvals. The condition is satisfied then the
+	// actual exit code value is in any of the intervals
+	// +optional
+	in      *[]ExitCodeInterval
+
+	// specifies the set of internvals. The condition is satisfied then the
+	// actual exit code value is not in any of the intervals
+	// +optional
+	notIn   *[]ExitCodeInterval
+
+	// Restricts the check for exit codes to only apply to the container with
+	// the specified name
+	// +optional
+	ContainerName *string
+}
+
+// PodStatusReasonCondition specifies the requirements for pod's status.reason
+type PodStatusReasonCondition struct {
+
+	// specifies the set of string values. The condition is satisfied when the
+	// actual value of pod's status reason is in this set
+	// +optional
+	in *[]string
+
+	// specifies the set of string values. The condition is satisfied when the
+	// actual value of pod's status reason is not in this set
+	// +optional
+	notIn *[]string
+}
+
+// BackoffPolicyRule specifies a backoff policy rule, which comprises an action
+// taken on a pod failure and the condition which needs to be satisfied to take
+// the action.
+type BackoffPolicyRule struct {
```

Comment:
I am afraid having new operator names like InClosedRange/NotInClosedRange will add some confusion, users may try to use them in labelSelector for example. Are they really necessary?

Also, it seems we have some canonical definitions here which include gt and lt (it doesn't cover the range case though): https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/apimachinery/pkg/selection/operator.go#L23

Reply:
They are not really necessary. However, they are useful to make syntax less verbose in common cases. 

Say, you want to terminate the job if the container exit code is in range [1,12]. 

Using the lt, gt operators only:
`      
    requirements:
      - source: ExitCode
        containerName: main-job-container
        operator: gt
        Values:
        - 0
      - source: ExitCode
        containerName: main-job-container
        operator: lt
        Values:
        - 13
`
(which means also duplication of the containerName field) vs using the dedicated operator:
`
    requirements:
      - source: ExitCode
        containerName: main-job-container
        operator: InClosedRange
        Values:
        - 1
        - 12
`

Reply:
Update in the latest proposal I drop the idea of using operators for restricting exit codes. Reasons:
- they might be misleading anyway for users of label selectors as they don't use the key field
- their use is more verbose when you want to express a relatively simple condition: fail the job if the exit code for the "main-job-container" is in range [1,10] or [100, 127]. This can be achieved in the latest API:

```yaml
 backoffPolicy:
   rules:
   - action: Terminate
     requirements:
     - onExitCodes:
         containerName: main-container
         inRange:
         - minInclusive: 1
           maxInclusive: 10
         - minInclusive: 100
           maxInclusive: 127
```

Reply:
The API is updated - please refer to KEP for the current version. Some alternatives are discussed there too.